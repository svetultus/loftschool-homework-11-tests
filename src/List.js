import React, { useState, useReducer } from "react";

const ADD_ITEM = "add_item"
const REMOVE_ITEM = "remove_item"

const addItem = (item) => ({type: ADD_ITEM, payload: item})
const removeItem = (item) => ({type: REMOVE_ITEM, payload: item})

const reducer = (state, action) => {
  switch (action.type) {
    case ADD_ITEM:
      return [...state, action.payload];
    case REMOVE_ITEM:
      return state.filter(item => item !== action.payload);
    default:
      throw new Error();
  }
};

export const List = () => {
  const [state, dispatch] = useReducer(reducer, []);
  const [text, setText] = useState("");

  const updateText = event => {
    setText(event.target.value);
  };

  const onAdd = () => {
    if (text && !state.includes(text)) {
      dispatch(addItem(text));
    }
    setText('');
  };

  const onRemove = value => () => {
    dispatch(removeItem(value));
  };

  return (
    <>
      <ul>
        {state.map(item => (
          <li key={item} onClick={onRemove(item)}>
            {item}
          </li>
        ))}
      </ul>
      <input value={text} onChange={updateText} />
      <button onClick={onAdd}>Add item</button>
    </>
  );
};
