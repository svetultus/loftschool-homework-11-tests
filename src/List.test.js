// Компонент List - список
// Напишите для него тесты.
// Компонент позволяет добавлять элементы в список.
// При добавлении нового элемента тексовое поле очищается.

// При клике на элементы списка они удаляются.

// Убедитесь, что вы протестировали всю функциональнось.

import React from "react";
import ReactDOM from "react-dom";
import List from "./List";
import renderer from "react-test-renderer";
import { render, fireEvent } from "@testing-library/react";
const domTestingLib = require("@testing-library/dom");
const { queryHelpers } = domTestingLib;
const queryByTagName = (...args) => queryHelpers.queryAllByTagName(...args);

it("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<List />, div);
  ReactDOM.unmountComponentAtNode(div);
});

describe("List", () => {
  it("соответсвует snapshot'у", () => {
    expect(renderer.create(<List />)).toMatchSnapshot();
  });

  it("добавляет значение input в List", () => {
    const wrapper = render(<List />);

    fireEvent.change(wrapper.queryByTestId("input"), {
      target: { value: "1000 чертей" }
    });
    fireEvent.click(wrapper.queryByTestId("button"));
    expect(wrapper.queryByText("1000 чертей")).toBeTruthy;
    expect(wrapper.queryByTestId("list").children.length).toBe(1);
    expect(wrapper.queryByTestId("input").value).toBe("");
    fireEvent.change(wrapper.queryByTestId("input"), {
      target: { value: "2000 чертей" }
    });
    fireEvent.click(wrapper.queryByTestId("button"));
    expect(wrapper.queryByText("2000 чертей")).toBeTruthy;
    expect(wrapper.queryByTestId("input").value).toBe("");
  });

  it("не добавляет пустое значение input в List", () => {
    const wrapper = render(<List />);

    fireEvent.change(wrapper.queryByTestId("input"), {
      target: { value: "" }
    });
    fireEvent.click(wrapper.queryByTestId("button"));
    expect(wrapper.queryByTestId("list").children.length).toBe(0);
  });

  it("удаляет элемент li при клике", () => {
    const wrapper = render(<List />);
    fireEvent.change(wrapper.queryByTestId("input"), {
      target: { value: "1000 чертей" }
    });
    fireEvent.click(wrapper.queryByTestId("button"));
    fireEvent.click(wrapper.queryByText("1000 чертей"));
    expect(wrapper.queryByText("1000 чертей")).toBeFalsy();
  });
});
