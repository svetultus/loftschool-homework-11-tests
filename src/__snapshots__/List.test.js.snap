// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List соответсвует snapshot'у 1`] = `
Array [
  <ul
    data-testid="list"
  />,
  <input
    data-testid="input"
    onChange={[Function]}
    value=""
  />,
  <button
    data-testid="button"
    onClick={[Function]}
  >
    Add item
  </button>,
]
`;
